core/parser.mly

```diff
@@ -929,22 +929,32 @@ forall_datatype:
      S = !(ModuleA.ModuleB.Type).S
 
      Parenthesised versions take priority over non-parenthesised versions.
+
+   jcheney: Modified grammar to allow unparenthesized qualified type names
+   in most places, but specifically forbid in ? . or ! .
+   This also requires moving the TILDE production into primary_datatype
+   since otherwise, ? ~M.x . y is ambiguous.
+   This is not ideal since it spreads the session-related constructs
+   among several nonterminals.
 */
 session_datatype:
-| BANG datatype DOT datatype                                   { `Output ($2, $4) }
-| QUESTION datatype DOT datatype                               { `Input  ($2, $4) }
+| BANG primary_datatype_loc DOT datatype                       { `Output ($2, $4) }
+| QUESTION primary_datatype_loc DOT datatype                   { `Input  ($2, $4) }
 | LBRACKETPLUSBAR row BARPLUSRBRACKET                          { `Select $2       }
 | LBRACKETAMPBAR row BARAMPRBRACKET                            { `Choice $2       }
-| TILDE datatype                                               { `Dual $2         }
 | END                                                          { `End             }
 | primary_datatype                                             { $1               }
+| qualified_type_name                                          { `QualifiedTypeApplication ($1, []) }
 
 parenthesized_datatypes:
 | LPAREN RPAREN                                                { [] }
-| LPAREN qualified_type_name RPAREN                            { [with_pos $loc (`QualifiedTypeApplication ($2, []))] }
 | LPAREN datatypes RPAREN                                      { $2 }
 
+primary_datatype_loc:
+| primary_datatype                                             { with_pos $loc $1 }
+
 primary_datatype:
+| TILDE primary_datatype_loc                                   { `Dual $2 }
```

Comment:
IIUC this will allow tilde in patterns, because:

```
pattern:
| typed_pattern                         { $1 }
| typed_pattern COLON primary_datatype  { with_pos $loc (`HasType ($1, datatype (with_pos $loc($3) $3))) }
```

Is that OK? If no then it needs to be fixed. If yes, then it would make sense for the `pattern` production to use `primary_datatype_loc` instead of currrent `with_pos $loc($3) $3`.

Reply:
I don't know that part of the grammar well, but it looks like the effect of my change would be that parentheses are not required in a pattern of the form 
```
typed_pattern COLON TILDE primary_datatype_loc
```
whereas in the current grammar one would have to parenthesize:
```
typed_pattern COLON LPAREN TILDE primary_datatype_loc RPAREN
```

However, I don't have a clear idea whether either syntax leads to something meaningful, i.e. whether typechecking/kindchecking ever succeeds for programs containing either syntactic form.  I suspect not, but either way the effect on the grammar seems to be to make it more liberal in terms of parentheses without changing the set of possible abstract syntax trees that can be represented.

Reply:
OK.  I only introduced it because I didn't know you could do $loc($3) to get the location info of a specific RHS nonterminal though, so we could also inline it in both places.

I also wonder if it would help to establish a convention that `nt_pos` means nonterminal `nt` with position information (e.g. `datatype` should be `datatype_pos`) or the reverse, say `nt` means with position information and `nt_node` means without.  The fact that different "type" nonterminals sometimes have position information and sometimes don't led to confusing misunderstandings on my part (and the resulting type error messages were not that helpful).  

Alternatively, probably enough to document the types in the mly file.

Reply:
> I also wonder if it would help to establish a convention that nt_pos means nonterminal nt with position information

I was wondering the same thing. I guess the only way to find out is trying it and seeing whether it improves readability of the code. BTW. most smart constructors in `SugarConstructors` produce values with positions, but on several occasions I have found the need to have constructors just for the nodes. These have names ending with `_node` suffic.